core/src/main/java/com/linecorp/armeria/client/limit/AsyncConcurrencyLimit.java

```diff
@@ -0,0 +1,230 @@
+        // maxPendingAcquires we may be able to run some pending tasks first and so allow to add
+        // more.
+        runTaskQueue();
+    }
+
+    private void runTaskQueue() {
+        while (acquiredPermitCount.get() < maxParallelism) {
+            final AcquireTask task = pendingAcquireQueue.poll();
+            if (task == null) {
+                break;
+            }
+
+            // Cancel the timeout if one was scheduled
+            final ScheduledFuture<?> timeoutFuture = task.timeoutFuture;
+            timeoutFuture.cancel(false);
+
+            --pendingAcquireCount;
+            task.acquired();
+
+            newPermit(task.promise);
+        }
+
+        // We should never have a negative value.
+        assert pendingAcquireCount >= 0;
+        assert acquiredPermitCount.get() >= 0;
+    }
+
+    private Future<SafeCloseable> newPermit(Promise<SafeCloseable> promise) {
+        promise.setSuccess(semaphorePermit);
+        return promise;
+    }
+
+    private final class AcquireTask implements FutureListener<SafeCloseable> {
+        final Promise<SafeCloseable> promise;
+        final long expireNanoTime = System.nanoTime() + acquireTimeoutNanos;
+        private final EventLoop executor;
+        final ScheduledFuture<?> timeoutFuture;
+        private boolean acquired;
+
+        AcquireTask(EventLoop executor, Promise<SafeCloseable> promise, ScheduledFuture<?> timeoutFuture) {
+            this.executor = executor;
+            this.timeoutFuture = timeoutFuture;
+            this.promise = promise.addListener(this);
+        }
+
+        @Override
+        public void operationComplete(Future<SafeCloseable> future) throws Exception {
+            assert executor.inEventLoop();
+
+            if (future.isSuccess()) {
+                promise.setSuccess(future.getNow());
+            } else {
+                if (acquired) {
+                    decrementAndRunTaskQueue();
+                } else {
+                    runTaskQueue();
+                }
+
+                promise.setFailure(future.cause());
+            }
+        }
+
+        public void acquired() {
+            if (acquired) {
+                return;
+            }
+            acquiredPermitCount.incrementAndGet();
+            acquired = true;
+        }
+    }
+
+    private class TimeoutTask implements Runnable {
+        @Override
+        public final void run() {
+            final long nanoTime = System.nanoTime();
+            for (;;) {
+                final AcquireTask task = pendingAcquireQueue.peek();
+                if (task == null || nanoTime - task.expireNanoTime < 0) {
+                    break;
+                }
+                pendingAcquireQueue.remove();
+
+                --pendingAcquireCount;
+                task.acquired();
+                task.promise.setFailure(UnprocessedRequestException.of(ConcurrencyLimitTimeoutException.get()));
```

Comment:
- I find this is inefficient. It'd be better just failing the future without incrementing and then immediately decrementing the counter.
- We could avoid instantiating exceptions if the future is complete already.